src/assisted_installer_controller/assisted_installer_controller.go

```diff
@@ -247,6 +247,168 @@ func (c controller) UpdateBMHs(wg *sync.WaitGroup) {
+			continue
+		}
+
+		_, isWorker := node.Labels["node-role.kubernetes.io/worker"]
+		if !(isWorker && bmh.Spec.ConsumerRef == nil) {
+			continue
+		}
+
+		allUpdated = false
+		var machine *mapiv1beta1.Machine
+		machine, err = c.findMachineFromNode(node, machineList)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to find machine from node %s", node.Name)
+			continue
+		}
+
+		// Set internalIP to same as node's as this will cause the
+		// machine operator's nodelink controller to link the node and machine:
+		// https://github.com/openshift/machine-api-operator/blob/master/pkg/controller/nodelink/nodelink_controller.go
+		// Note: while this seems like a good idea, it causes worker machines to error:
+		// errorMessage: Can't find created instance.
+		// this is because of:
+		// https://github.com/openshift/machine-api-operator/blob/master/pkg/controller/machine/controller.go#L304-L309
+		/*
+			if  len(machine.Status.Addresses) == 0 {
+				machine.Status.Addresses = []v1.NodeAddress{*interalAddress}
+				log.Infof("updating internal address")
+
+				err = c.kc.UpdateMachineStatus(machine)
+				if err != nil {
+					log.WithError(err).Errorf("Failed to update machine internal address %s", machine.Name)
+					continue
+				}
+			}
+		*/
+		bmh.Spec.ConsumerRef = &v1.ObjectReference{
+			APIVersion: machine.APIVersion,
+			Kind:       machine.Kind,
+			Namespace:  machine.Namespace,
+			Name:       machine.Name,
+		}
+		log.Infof("updating bmh consumerRef")
+		err = c.kc.UpdateBMH(&bmh)
+		if err != nil {
+			log.WithError(err).Errorf("Failed to remove status annotation from BMH %s", bmh.Name)
+		}
+	}
+	return allUpdated
+}
+
+func (c controller) findNodeFromBMH(bmh *metal3v1alpha1.BareMetalHost, nodeList *v1.NodeList) (*v1.Node, *v1.NodeAddress, error) {
+	if bmh.Status.HardwareDetails == nil {
+		return nil, nil, fmt.Errorf("BMH %s hardware details not populated", bmh.Name)
+	}
+	for _, node := range nodeList.Items {
+		for _, a := range node.Status.Addresses {
+			if a.Type != v1.NodeInternalIP {
+				continue
+			}
+			for _, nic := range bmh.Status.HardwareDetails.NIC {
+				//	nics:
+				//	- ip: 192.168.126.13/24
+				//  addresses:
+				//  - address: 192.168.126.13
+				//    type: InternalIP
+
+				if strings.HasPrefix(nic.IP, a.Address) {
+					return &node, &a, nil
+				}
+			}
+		}
+	}
+
+	return nil, nil, fmt.Errorf("can not find node from BMH %v", bmh.Name)
+}
+
+func (c controller) findMachineFromNode(node *v1.Node, machineList *mapiv1beta1.MachineList) (*mapiv1beta1.Machine, error) {
+	var internalAddress v1.NodeAddress
+	for _, a := range node.Status.Addresses {
+		if a.Type == v1.NodeInternalIP {
+			internalAddress = a
+		}
+	}
+	if internalAddress.Address == "" {
+		return nil, fmt.Errorf("can not find internal address of node %v", node.Name)
+	}
+
+	var firstUnassociatedWorker *mapiv1beta1.Machine
+	for i, machine := range machineList.Items {
```

Comment:
i think you should keep a list of  all the machines that were already mapped or remove a mapped machine from the list. Otherwise, in case Actuator has not succeeded to update Machine status ip address before your loop moves to the next BMH, you can pick the same machine.